### PR TITLE
CodeMirror blob: Rewrite hovercard logic

### PR DESCRIPTION
**Pull request by fkling**
This is a rewrite of the hovercard rendering logic to improve its usability.

**tl;dr**: It's easier to move the mouse to the hovercard without it disappearing and the hovercard now properly aligns with the token range returned by the server.

The new implementation should be much easier to understand because data flows primarily in one direction whereas in the old implementation there was a lot of back and forth between various components.

Lets start from the end:

"Rendering a hovercard" involves the following components:
- Showing a CodeMirror tooltip (the actual hovercard)
- Decorating the corresponding token to highlight it
- Adding DOM event handlers to handle "click-to-go-to-definition"

All these are extensions provided by CodeMirror and are enabled via the `showHovercard` facet, which holds a list of hovercards to render.

This facet gets its input from the `hovercardState` field, which keeps track of the currently pinned hovercard and the hovercard of the currently hovered token.

The field gets update by two managers: `pinManager` and `hoverManager`. Both of these take inputs (the pin manager the location derived from the URL; the hover manager mouse move events), determine the token at the provided position and query the hover provider for information. Only when that provider returns useable information do these managers create `Hovercard`s and update the field (that's the biggest change to the previous implementation where we created "empty" tooltips and let the hovercard view implementation query the hover provider).

----

Some additional changes:

- As long as the mouse cursor doesn't move *away* from the currently opened hovercard *and* there is no new hover information of the position under the cursor, the hovercard stays open. That should make it easier to reach the hovercard.
- Debouncing mouse events is still necessary to not immediately open new hovercards while potentially moving the mouse cursor over other tokens.
- I added a simple cache for the hover information to prevent flickers and unnecessary data requests when hovering over previously hovered tokens. The cache clears when the document changes.
- I removed the diagram because I think the data flow is clearer now.
- I added a couple of additional tests to verify core hovercard behavior.

## Test plan
Manual testing and new integration tests.

## App preview:

- [Web](https://sg-web-fkling-cm-hovercard-behavior.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-tgzmpstozs.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Review by olafurpg — APPROVED** [review]
I just tested this locally and the new popover behavior is a dramatic upgrade over before
